src/main/java/bio/terra/service/filedata/DrsService.java

```diff
@@ -146,69 +180,122 @@ public void close() {
   public DRSObject lookupObjectByDrsId(
       AuthenticatedUserRequest authUser, String drsObjectId, Boolean expand) {
     try (DrsRequestResource r = new DrsRequestResource()) {
-      DrsId drsId = drsIdService.fromObjectId(drsObjectId);
-      SnapshotProject snapshotProject;
-      try {
-        UUID snapshotId = UUID.fromString(drsId.getSnapshotId());
-        // We only look up DRS ids for unlocked snapshots.
-        String retrieveTimer = performanceLogger.timerStart();
-
-        snapshotProject = snapshotService.retrieveAvailableSnapshotProject(snapshotId);
-
-        performanceLogger.timerEndAndLog(
-            retrieveTimer,
-            drsObjectId, // not a flight, so no job id
-            this.getClass().getName(),
-            "snapshotService.retrieveAvailable");
-      } catch (IllegalArgumentException ex) {
-        throw new InvalidDrsIdException("Invalid object id format '" + drsObjectId + "'", ex);
-      } catch (SnapshotNotFoundException ex) {
-        throw new DrsObjectNotFoundException(
-            "No snapshot found for DRS object id '" + drsObjectId + "'", ex);
-      }
+      Snapshot snapshot = lookupSnapshotForDRSObject(drsObjectId);
 
-      // Make sure requester is a READER on the snapshot
       String samTimer = performanceLogger.timerStart();
-
       samService.verifyAuthorization(
-          authUser, IamResourceType.DATASNAPSHOT, drsId.getSnapshotId(), IamAction.READ_DATA);
-
+          authUser, IamResourceType.DATASNAPSHOT, snapshot.getId().toString(), IamAction.READ_DATA);
       performanceLogger.timerEndAndLog(
           samTimer,
           drsObjectId, // not a flight, so no job id
           this.getClass().getName(),
           "samService.verifyAuthorization");
 
-      int depth = (expand ? -1 : 1);
-
-      FSItem fsObject;
-      try {
-        String lookupTimer = performanceLogger.timerStart();
-        fsObject = fileService.lookupSnapshotFSItem(snapshotProject, drsId.getFsObjectId(), depth);
-
-        performanceLogger.timerEndAndLog(
-            lookupTimer,
-            drsObjectId, // not a flight, so no job id
-            this.getClass().getName(),
-            "fileService.lookupSnapshotFSItem");
-      } catch (InterruptedException ex) {
-        throw new FileSystemExecutionException(
-            "Unexpected interruption during file system processing", ex);
-      }
+      return lookupDRSObjectAfterAuth(expand, snapshot, drsObjectId, authUser, false);
+    }
+  }
 
-      if (fsObject instanceof FSFile) {
-        return drsObjectFromFSFile((FSFile) fsObject, drsId.getSnapshotId(), authUser);
-      } else if (fsObject instanceof FSDir) {
-        return drsObjectFromFSDir((FSDir) fsObject, drsId.getSnapshotId());
-      }
+  @VisibleForTesting
+  Snapshot lookupSnapshotForDRSObject(String drsObjectId) {
```

Comment:
Given how big a snapshot is, I think we should probably avoid using the entire snapshot object if we can avoid it.  I think the `SnapshotCacheResult` object in #1253 might be worthwhile here

Reply:
Definitely! Once that PR goes in, I'll add the needed fields to the cache result.

Reply:
Should be added now!